torch/csrc/jit/codegen/cuda/lower_index_hoist.cpp

```diff
@@ -12,419 +12,275 @@ namespace cuda {
+    auto loop = loops[level++];
+    if (!loop->isTrivial()) {
+      return true;
+    }
   }
+  return false;
+}
 
-  // Find used loops and their index vals
-  for (const auto i : c10::irange(loops.size())) {
-    auto loop = loops.at(i);
-    auto loop_id = gpu_lower->caMap()->getConcreteMappedID(
-        loop->iter_domain(), IdMappingMode::LOOP);
-    auto it = concrete_leaf_ids.find(loop_id);
-    if (it != concrete_leaf_ids.end()) {
-      // This leaf reference id is used for indexing the consumer id
-      used_loops_.push_back(loop);
-      auto index_it = ref_index_map.find(ref_td->axis(i));
-      TORCH_INTERNAL_ASSERT(
-          index_it != ref_index_map.end(),
-          "Index not found for leaf ID, ",
-          ref_td->axis(i)->toString());
-      loop_index_vals_.push_back(index_it->second);
+// Check if in the definition of from, there is a subexpression equivalent to
+// reference. If found, then return this subexpression.
+Val* findCommonSubexpr(Val* from, Val* reference) {
+  if (from->sameAs(reference)) {
+    return from;
+  }
+  auto def = from->definition();
+  if (def != nullptr) {
+    for (auto input : def->inputs()) {
+      auto common_subexpr = findCommonSubexpr(input, reference);
+      if (common_subexpr != nullptr) {
+        return common_subexpr;
+      }
     }
   }
-
-  TORCH_INTERNAL_ASSERT(
-      !used_loops_.empty(),
-      "No loop used for indexing found. ",
-      consumer_indexed_id->toString());
-
-  TORCH_INTERNAL_ASSERT(
-      consumer_leaf_ids.size() == used_loops_.size(),
-      "consumer_leaf_ids.size() = ",
-      consumer_leaf_ids.size(),
-      ", used_loops_.size() == ",
-      used_loops_.size(),
-      ", loops.size() == ",
-      loops.size());
+  return nullptr;
 }
 
-CommonIndexKey::CommonIndexKey(
-    IterDomain* consumer_indexed_id,
-    TensorDomain* consumer_td,
-    const std::vector<IterDomain*>& loop_domains,
-    const std::unordered_map<IterDomain*, Val*>& loop_index_map,
+// Find which loop variables `index` depend on
+std::unordered_set<Val*> getLoopVariableDependency(
+    Val* index,
     const std::vector<kir::ForLoop*>& loops) {
-  auto gpu_lower = GpuLower::current();
-
-  concrete_indexed_id_ = gpu_lower->caMap()->getConcreteMappedID(
-      consumer_indexed_id, IdMappingMode::EXACT);
-
-  const auto consumer_leaf_ids =
-      getUsedLeafIds(consumer_indexed_id, consumer_td);
-
-  // Convert to Parallel concrete IDs to find matching loops.
-  std::unordered_set<IterDomain*> concrete_leaf_ids;
-  for (auto& id : consumer_leaf_ids) {
-    concrete_leaf_ids.insert(
-        gpu_lower->caMap()->getConcreteMappedID(id, IdMappingMode::LOOP));
+  auto def = index->definition();
+
+  // If `index` is not computed from other values, then it must either be a loop
+  // variable or something like named scalar or constant
+  if (def == nullptr) {
+    // Check if `index` is a loop variable
+    for (auto loop : loops) {
+      if (loop->isTrivial()) {
```

Comment:
added:
```C++
      // We skip trivial loop here because it is never materialized, so its loop
      // variable is accessible everywhere. For example, if the trivial loop is
      // a un-concretized broadcast, the its loop variable is a constant 0. If
      // the trivial loop is thread parallelized, then its loop variable will be
      // threadIdx.x, which is also accessible on all scopes.
```